src/sql/workbench/contrib/notebook/browser/markdownToolbarActions.ts

```diff
@@ -0,0 +1,241 @@
+						}
+						operations.push({ range: endRange, text: endInsertedCode });
+						editorModel.pushEditOperations(selections, operations, null);
+					}
+				}
+				this.setEndSelection(endRange, type, editorControl, nothingSelected);
+			}
+		}
+	}
+
+	// For items like lists (where we need to insert a character at the beginning of each line), create
+	// range object for that range
+	private transformRangeByLineOffset(range: IRange, lineOffset: number): IRange {
+		return {
+			startColumn: lineOffset === 0 ? range.startColumn : 1,
+			endColumn: range.endColumn,
+			startLineNumber: range.endLineNumber + lineOffset,
+			endLineNumber: range.endLineNumber + lineOffset
+		};
+	}
+
+	private getStartTextToInsert(type: MarkdownButtonType): string {
+		switch (type) {
+			case MarkdownButtonType.BOLD:
+				return '**';
+			case MarkdownButtonType.ITALIC:
+				return '_';
+			case MarkdownButtonType.CODE:
+				return '```\n';
+			case MarkdownButtonType.LINK:
+				return '[';
+			case MarkdownButtonType.UNORDERED_LIST:
+				return '- ';
+			case MarkdownButtonType.ORDERED_LIST:
+				return '1. ';
+			case MarkdownButtonType.IMAGE:
+				return '![ALT TEXT](';
+			case MarkdownButtonType.HIGHLIGHT:
+				return '<mark>';
+			default:
+				return '';
+		}
+	}
+
+	private getEndTextToInsert(type: MarkdownButtonType): string {
+		switch (type) {
+			case MarkdownButtonType.BOLD:
+				return '**';
+			case MarkdownButtonType.ITALIC:
+				return '_';
+			case MarkdownButtonType.CODE:
+				return '\n```';
+			case MarkdownButtonType.LINK:
+				return ']()';
+			case MarkdownButtonType.IMAGE:
+				return ')';
+			case MarkdownButtonType.HIGHLIGHT:
+				return '</mark>';
+			case MarkdownButtonType.UNORDERED_LIST:
+			case MarkdownButtonType.ORDERED_LIST:
+			default:
+				return '';
+		}
+	}
+
+	private getMarkdownLineType(type: MarkdownButtonType): MarkdownLineType {
+		switch (type) {
+			case MarkdownButtonType.UNORDERED_LIST:
+			case MarkdownButtonType.ORDERED_LIST:
+				return MarkdownLineType.EVERY_LINE;
+			default:
+				return MarkdownLineType.BEGIN_AND_END_LINES;
+		}
+	}
+
+	// Get offset from the end column for editor selection
+	// For example, when inserting a link, we want to have the cursor be present in between the brackets
+	private getColumnOffsetForSelection(type: MarkdownButtonType, nothingSelected: boolean): number {
+		if (nothingSelected) {
+			return 0;
+		}
+		switch (type) {
+			case MarkdownButtonType.LINK:
+				return 2;
+			case MarkdownButtonType.IMAGE:
+				return 0;
+			// -1 is considered as having no explicit offset, so do not do anything with selection
+			default: return -1;
+		}
+	}
+
+	private getEditorControl(): CodeEditorWidget {
+		if (!this._notebookEditor) {
+			this._notebookEditor = this._notebookService.findNotebookEditor(this._cellModel.notebookModel.notebookUri);
+		}
+		if (this._notebookEditor && this._notebookEditor.cellEditors && this._notebookEditor.cellEditors.length > 0) {
```

Comment:
nit: You can shorten this by just doing `this._notebookEditor?.cellEditors?.length > 0`

Reply:
I keep forgetting this syntax exists now. Thank you. Definitely much cleaner.

Reply:
That's some sweet implied null checking there!